src/formats/DCD.cpp

```diff
@@ -0,0 +1,690 @@
+
+    auto compute_frame_size = [&](size_t n_atoms) {
+        size_t result = 0;
+
+        size_t marker_size = options_.use_64_bit_markers ? 8 : 4;
+        if (options_.charmm_format && options_.charmm_unitcell) {
+            // 6 doubles and two markers for the unit cell
+            result += 2 * marker_size + 6 * sizeof(double);
+        }
+        // three coordinates sets, each with two markers
+        result += 3 * (2 * marker_size + n_atoms * sizeof(float));
+        if (options_.has_4d_data) {
+            // optional fourth dimension
+            result += 2 * marker_size + n_atoms * sizeof(float);
+        }
+
+        return result;
+    };
+
+    first_frame_size_ = compute_frame_size(n_atoms_);
+    if (fixed_atoms_.empty()) {
+        frame_size_ = first_frame_size_;
+    } else {
+        frame_size_ = compute_frame_size(n_free_atoms_);
+    }
+
+    auto file_size = file_->file_size();
+    auto n_frames_from_size = (file_size - header_size_ - first_frame_size_) / frame_size_ + 1;
+    if (n_frames_from_size != n_frames_) {
+        warning(
+            "DCD reader",
+            "the file header claims {} frames, but the file size indicate we have {}",
+            n_frames_,
+            n_frames_from_size
+        );
+        n_frames_ = static_cast<size_t>(n_frames_from_size);
+    }
+
+    if (!fixed_atoms_.empty()) {
+        this->read_fixed_coordinates();
+    }
+}
+
+UnitCell DCDFormat::read_cell() {
+    if (!options_.charmm_format || !options_.charmm_unitcell) {
+        return {};
+    }
+
+    this->expect_marker(6 * sizeof(double));
+    double buffer[6] = {0};
+    file_->read_f64(buffer, 6);
+    this->expect_marker(6 * sizeof(double));
+
+    if (options_.charmm_format && options_.charmm_version > 25) {
+        // recent charmm version store cell vectors directly. It seems to be
+        // doing this since CHARMM 22, but the readdcd code pretends to be
+        // CHARMM 24 most of the time, so we need to check for a higher version.
+        auto matrix = Matrix3D(
+            buffer[0], buffer[1], buffer[3],
+            buffer[1], buffer[2], buffer[4],
+            buffer[3], buffer[4], buffer[5]
+        );
+        return UnitCell(matrix);
+    }
+
+    auto lengths = Vector3D(buffer[0], buffer[2], buffer[5]);
+    auto alpha = buffer[4];
+    auto beta = buffer[3];
+    auto gamma = buffer[1];
+
+    if (std::fabs(alpha) <= 1.0 && std::fabs(beta) <= 1.0 && std::fabs(gamma) <= 1.0) {
+        // cell angles are either stored directly or as the cos of the angle
+        alpha = cos_to_angle_degrees(alpha);
+        beta = cos_to_angle_degrees(beta);
+        gamma = cos_to_angle_degrees(gamma);
+    }
+
+    auto is_degree_angle = [](double angle) {
+        return angle >= 0.0 && angle <= 180.0;
+    };
+
+    if (
+        lengths[0] >= 0.0 && lengths[1] >= 0.0 && lengths[2] >= 0.0 &&
+        is_degree_angle(alpha) && is_degree_angle(beta) && is_degree_angle(gamma)
+    ) {
```

Comment:
It could be, but I find this version clearer: we are trying to determine how the cell is stored in the file between the different options. I'll add a comment to clarify

Reply:
Good point! An explicit check made sense in a previous version of this function, but it no longer required!

Reply:
Before [53dbace](https://github.com/chemfiles/chemfiles/pull/453/commits/53dbace80d408c10bd80f57e3f4843cbdb0a6f9f#diff-5309e1bd6661c52b53d2750bc766f5ba9136a0407c9cb528c7770e69eabfc5fdL392-R391), a warning was emitted when a UnitCell could not be constructed (e.g. due to negative angles). Now, the constructor simply throws.
Is this a wanted change or should there be a catch block in case the construction fails?

Reply:
I think that having a cell we are fully unable to parse should not happen that often, and it might be better to throw an error in this case, in the same way that an invalid cell in a XYZ or GRO file would throw an error.